From be0efde6b3f7586b595867d2eb54e750c862f8f2 Mon Sep 17 00:00:00 2001
From: Tom Marshall <tdm.code@gmail.com>
Date: Wed, 23 Aug 2017 22:11:03 +0000
Subject: [PATCH 23/26] adb: Look for executables in alternate places

 * Restore global variable recovery_mode.  This is set based on the
   commandline parameter --banner.

 * Use /sbin for path to 'sh' and 'bu' when in recovery mode.

 * Honor persist.sys.adb.shell when not in recovery mode.

Change-Id: I02ac651397021097fe865436c982e2b720d7917f
---
 adb/adb.h                  |  4 ++++
 adb/daemon/main.cpp        |  4 ++++
 adb/services.cpp           | 15 ++++++++++++---
 adb/shell_service.cpp      | 28 +++++++++++++++++++++++++---
 adb/shell_service_test.cpp |  2 ++
 5 files changed, 47 insertions(+), 6 deletions(-)

diff --git a/adb/adb.h b/adb/adb.h
index 1b2e4aa..0b75517 100644
--- a/adb/adb.h
+++ b/adb/adb.h
@@ -199,6 +199,10 @@ int  local_connect_arbitrary_ports(int console_port, int adb_port, std::string*
 
 ConnectionState connection_state(atransport *t);
 
+#if !ADB_HOST
+extern int recovery_mode;
+#endif
+
 extern const char* adb_device_banner;
 
 #if !ADB_HOST
diff --git a/adb/daemon/main.cpp b/adb/daemon/main.cpp
index 52716e3..4f7b3c5 100644
--- a/adb/daemon/main.cpp
+++ b/adb/daemon/main.cpp
@@ -211,6 +211,8 @@ int adbd_main(int server_port) {
     return 0;
 }
 
+int recovery_mode = 0;
+
 int main(int argc, char** argv) {
     while (true) {
         static struct option opts[] = {
@@ -243,6 +245,8 @@ int main(int argc, char** argv) {
         }
     }
 
+    recovery_mode = (strcmp(adb_device_banner, "recovery") == 0);
+
     close_stdin();
 
     debuggerd_init(nullptr);
diff --git a/adb/services.cpp b/adb/services.cpp
index c9ea039..6ad49ea 100644
--- a/adb/services.cpp
+++ b/adb/services.cpp
@@ -271,6 +271,13 @@ static int create_service_thread(void (*func)(int, void *), void *cookie)
     return s[0];
 }
 
+#if !ADB_HOST
+static const char* bu_path()
+{
+    return (recovery_mode ? "/sbin/bu" : "/system/bin/bu");
+}
+#endif
+
 int service_to_fd(const char* name, const atransport* transport) {
     int ret = -1;
 
@@ -304,12 +311,14 @@ int service_to_fd(const char* name, const atransport* transport) {
     } else if(!strncmp(name, "unroot:", 7)) {
         ret = create_service_thread(restart_unroot_service, NULL);
     } else if(!strncmp(name, "backup:", 7)) {
-        ret = StartSubprocess(android::base::StringPrintf("/system/bin/bu backup %s",
+        ret = StartSubprocess(android::base::StringPrintf("%s backup %s",
+                                                          bu_path(),
                                                           (name + 7)).c_str(),
                               nullptr, SubprocessType::kRaw, SubprocessProtocol::kNone);
     } else if(!strncmp(name, "restore:", 8)) {
-        ret = StartSubprocess("/system/bin/bu restore", nullptr, SubprocessType::kRaw,
-                              SubprocessProtocol::kNone);
+        ret = StartSubprocess(android::base::StringPrintf("%s restore",
+                                                         bu_path()).c_str(),
+                              nullptr, SubprocessType::kRaw, SubprocessProtocol::kNone);
     } else if(!strncmp(name, "tcpip:", 6)) {
         int port;
         if (sscanf(name + 6, "%d", &port) != 1) {
diff --git a/adb/shell_service.cpp b/adb/shell_service.cpp
index d4f334b..4d088b7 100644
--- a/adb/shell_service.cpp
+++ b/adb/shell_service.cpp
@@ -86,6 +86,7 @@
 #include <pty.h>
 #include <pwd.h>
 #include <sys/select.h>
+#include <sys/stat.h>
 #include <termios.h>
 
 #include <memory>
@@ -104,8 +105,26 @@
 #include "adb_utils.h"
 #include "security_log_tags.h"
 
+#include "cutils/properties.h"
+
 namespace {
 
+static std::string get_sh_path()
+{
+    if (recovery_mode) {
+        return "/sbin/sh";
+    }
+
+    char propbuf[PROPERTY_VALUE_MAX];
+    struct stat st;
+
+    property_get("persist.sys.adb.shell", propbuf, "");
+    if (propbuf[0] && stat(propbuf, &st) == 0) {
+        return propbuf;
+    }
+    return _PATH_BSHELL;
+}
+
 // Reads from |fd| until close or failure.
 std::string ReadAll(int fd) {
     char buffer[512];
@@ -324,12 +343,15 @@ bool Subprocess::ForkAndExec(std::string* error) {
         // processes, so we need to manually reset back to SIG_DFL here (http://b/35209888).
         signal(SIGPIPE, SIG_DFL);
 
+        std::string sh_path = get_sh_path();
+
         if (command_.empty()) {
-            execle(_PATH_BSHELL, _PATH_BSHELL, "-", nullptr, cenv.data());
+            execle(sh_path.c_str(), sh_path.c_str(), "-", nullptr, cenv.data());
         } else {
-            execle(_PATH_BSHELL, _PATH_BSHELL, "-c", command_.c_str(), nullptr, cenv.data());
+            execle(sh_path.c_str(), sh_path.c_str(), "-c", command_.c_str(), nullptr, cenv.data());
         }
-        WriteFdExactly(child_error_sfd, "exec '" _PATH_BSHELL "' failed: ");
+        std::string errmsg = "exec '" + sh_path + "' failed: ";
+        WriteFdExactly(child_error_sfd, errmsg.c_str());
         WriteFdExactly(child_error_sfd, strerror(errno));
         child_error_sfd.reset(-1);
         _Exit(1);
diff --git a/adb/shell_service_test.cpp b/adb/shell_service_test.cpp
index 839284e..645eb09 100644
--- a/adb/shell_service_test.cpp
+++ b/adb/shell_service_test.cpp
@@ -29,6 +29,8 @@
 #include "adb_io.h"
 #include "sysdeps.h"
 
+int recovery_mode = 0;
+
 class ShellServiceTest : public ::testing::Test {
   public:
     static void SetUpTestCase() {
-- 
2.7.4

